### DISCUSSION TITLE
Using jc-doyle/cmp-pandoc-references with blink

### DISCUSSION THREAD

**Discussion by manolomartinez** [LazyVim/LazyVim#5117] [Q&A] [answered]
Hi, I've been trying to use [cmp-pandoc-references](https://github.com/jc-doyle/cmp-pandoc-references) with the new blink.cmp, so far unsuccessfully. This is the relevant section of a config file called `pandoc.lua`:

    {
        "saghen/blink.compat",
        -- use the latest release, via version = '*', if you also use the latest release for blink.cmp
        optional = true,
        -- make sure to set opts so that lazy.nvim calls blink.compat's setup
        opts = {},
        version = not vim.g.lazyvim_blink_main and "*",
      },
    
      {
        "saghen/blink.cmp",
        dependencies = { "jc-doyle/cmp-pandoc-references" },
        providers = {
          -- create provider
          pandoc_references = {
            name = "pandoc_references", -- IMPORTANT: use the same name as you would for nvim-cmp
            module = "blink.compat.source",
          },
        },
        sources = {
          completion = { "pandoc_references" },
        },
      },
    

I'm a bit fumbling in the dark here, to be honest, but anyway now nvim complains with the following message:

    Failed to source `/home/manolo/.local/share/nvim/lazy/cmp-pandoc-references/after/plugin/cmp-pandoc-references.lua`
    
    vim/_editor.lua:0: /home/manolo/.config/nvim/init.lua..nvim_exec2() called at /home/manolo/.config/nvim/init.lua:0../home/manolo/.local/share/nvim/lazy/cmp-pandoc-references/after/plugin/cmp-pandoc-references.lua: Vim(source):E5113: Error while calling lua chunk: ...pandoc-references/after/plugin/cmp-pandoc-references.lua:1: module 'cmp' not found:
    ^Ino field package.preload['cmp']
    cache_loader: module cmp not found
    cache_loader_lib: module cmp not found
    ^Ino file './cmp.lua'
    ^Ino file '/usr/share/luajit-2.1/cmp.lua'
    ^Ino file '/usr/local/share/lua/5.1/cmp.lua'
    ^Ino file '/usr/local/share/lua/5.1/cmp/init.lua'
    ^Ino file '/usr/share/lua/5.1/cmp.lua'
    ^Ino file '/usr/share/lua/5.1/cmp/init.lua'
    ^Ino file './cmp.so'
    ^Ino file '/usr/local/lib/lua/5.1/cmp.so'
    ^Ino file '/usr/lib/lua/5.1/cmp.so'
    ^Ino file '/usr/local/lib/lua/5.1/loadall.so'
    stack traceback:
    ^I[C]: in function 'require'
    ^I...pandoc-references/after/plugin/cmp-pandoc-references.lua:1: in main chunk
    ^I[C]: in function 'nvim_exec2'
    ^Ivim/_editor.lua: in function 'cmd'
    ^I...local/share/nvim/lazy/lazy.nvim/lua/lazy/core/loader.lua:510: in function <...local/share/nvim/lazy/lazy.nvim/lua/lazy/core/loader.lua:509>
    ^I[C]: in function 'xpcall'
    ^I.../.local/share/nvim/lazy/lazy.nvim/lua/lazy/core/util.lua:135: in function 'try'
    ^I...local/share/nvim/lazy/lazy.nvim/lua/lazy/core/loader.lua:509: in function 'source'
    ^I...local/share/nvim/lazy/lazy.nvim/lua/lazy/core/loader.lua:457: in function 'source_runtime'
    ^I...local/share/nvim/lazy/lazy.nvim/lua/lazy/core/loader.lua:150: in function 'startup'
    ^I...anolo/.local/share/nvim/lazy/lazy.nvim/lua/lazy/init.lua:112: in function 'setup'
    ^I/home/manolo/.config/nvim/lua/config/lazy.lua:9: in main chunk
    ^I[C]: in function 'require'
    ^I/home/manolo/.config/nvim/init.lua:2: in main chunk
    
    # stacktrace:
      - vim/_editor.lua:0 _in_ **cmd**
      - ~/.config/nvim/lua/config/lazy.lua:9
      - ~/.config/nvim/init.lua:2
  
Any help appreciated!

**Comment by abeldekat** [comment]
The error suggest that blink.compat is not installed. That's because you copied `optional = true`.

This should work:

```lua
return { -- pandoc.lua
  {
    "saghen/blink.cmp",
    -- specify blink.compat as a dependency. Now its not optional anymore:
    dependencies = { "jc-doyle/cmp-pandoc-references", "saghen/blink.compat" },
    opts = {
      sources = {
        compat = { "pandoc_references" },
      },
    },
  },
}
```

In this case you don't even need to register a provider yourself: LazyVim will add the name/module automatically. See [here](https://github.com/LazyVim/LazyVim/blob/28d3ee49709978eb350e5c6bfe1f2ce36138c1d3/lua/lazyvim/plugins/extras/coding/blink.lua#L97).

**Reply by ultimatile** [reply]
This worked for me. You reminded me that I also needed to include "saghen/blink.compat" in the dependencies. Thank you.

**Reply by manolomartinez** [reply]
So, when you do the above you get a list of available references when you start tying `@...`, right?

**Reply by ultimatile** [reply]
I simply forgot to add the dependency to enable another source via blink.compat.
I didn't encounter the same issue you did. I apologize for the confusion. 

~I haven't used `pandoc_references` before. I've just tried the configuration above, and this is what happens when I type `@` in my bib file: Is this correct?~ Sorry, I've tried it in a completely wrong way...

**Reply by ultimatile** [reply]
I've retried it again with the above config, `@` did not trigger the completion, and `@` + some key triggered the completion.
Next, I've tried the config with `override`, and it seems to work:
![スクリーンショット 2025-01-18 4 50 45](https://github.com/user-attachments/assets/dcd1dfe3-2903-494b-ba2b-0f63b2c45be5)
Three items listed are valid item names prepared in my bibtex file.

**Reply by manolomartinez** [reply]
Thanks a lot. I'll try again if I can make it work :)

**Comment by manolomartinez** [comment]
Thanks for this! 

It doesn't quite work yet. What this source is supposed to do is complete bibtex entries in markdown. This works with nvim.comp, but not (yet) with blink: in the former typing an @ triggers the autocompletion, reading a bibtex file, as expected. In the latter nothing happens.

**Reply by abeldekat** [reply]
I am not using the plugin myself unfortunately. Perhaps this works?

```lua
return { -- pandoc.lua
  {
    "saghen/blink.cmp",
    -- specify blink.compat as a dependency. Now its not optional anymore:
    dependencies = {
      "jc-doyle/cmp-pandoc-references",
      "saghen/blink.compat",
    },
    opts = {
      sources = {
        compat = { "pandoc_references" },
        providers = {
          pandoc_references = {
            override = {
              get_trigger_characters = function()
                return { "@" }
              end,
            },
          },
        },
      },
    },
  },
}

```

**Comment by manolomartinez** [comment]
> I am not using the plugin myself unfortunately

Of course! Thanks a lot for all your help, @abeldekat . I'm sure I'll be able to fill in the gaps and make it work eventually :)

**Reply by abeldekat** [reply]
You're welcome!

I don't have a "bib" file but I did test the last code fragment using stock LazyVim. When I press "@", the cmp plugin sends back an empty table. Without the override, completion is not activated.

If the solution is correct, could you mark it as accepted answer for other people to use?

Best regards!

**Comment by manolomartinez** [comment]
Unfortunately not yet, no. But you've given me enough info that tinkering should take me the rest of the way there. I'll be sure to document my solution if and when I find it. Thanks again!

**Comment by a-torgovitsky** [comment, marked as answer]
Similar discussion in https://github.com/LazyVim/LazyVim/discussions/5102 in case it's helpful

**Comment by manolomartinez** [comment]
Yes! This finally does work, thanks a lot.

**Comment by frodew** [comment]
Copy-past solution for LazyVim for anyone wondering:

``` 
{
    "saghen/blink.cmp",
    dependencies = {
      "jc-doyle/cmp-pandoc-references",
      "saghen/blink.compat",
    },
    opts = {
      sources = {
        compat = { "pandoc_references" },
      },
    },
  },
```
